File: question2.c
#include <stdio.h>
#include <stdlib.h>

struct Node
{
    int data;
    struct Node *next;
};

struct Node *newNode(int data)
{
    struct Node *node = (struct Node *)malloc(sizeof(struct Node));
    node->data = data;
    node->next = NULL;
    return node;
}

struct Node *insertInMiddle(struct Node *head, int x)
{
    struct Node *new_node = newNode(x);

    if (head == NULL)
    {
        return new_node;
    }

    struct Node *slow = head, *fast = head;

    while (fast->next != NULL && fast->next->next != NULL)
    {
        slow = slow->next;
        fast = fast->next->next;
    }

    new_node->next = slow->next;
    slow->next = new_node;

    return head;
}

void printList(struct Node *head)
{
    struct Node *curr = head;
    while (curr != NULL)
    {
        printf("%d", curr->data);
        if (curr->next != NULL)
            printf(" ");
        curr = curr->next;
    }
    printf("\n");
}

int main()
{
    int x;
    struct Node *head = NULL, *tail = NULL;

    char buffer[100000];
    if (fgets(buffer, sizeof(buffer), stdin) == NULL)
    {

        head = NULL;
    }
    else
    {

        int val, i = 0;
        char *ptr = buffer;
        while (sscanf(ptr, "%d%n", &val, &i) == 1)
        {
            struct Node *node = newNode(val);
            if (head == NULL)
            {
                head = node;
                tail = node;
            }
            else
            {
                tail->next = node;
                tail = node;
            }
            ptr += i;
        }
    }

    scanf("%d", &x);

    head = insertInMiddle(head, x);

    printList(head);

    struct Node *temp;
    while (head != NULL)
    {
        temp = head;
        head = head->next;
        free(temp);
    }
}

File: question2.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def insertInMiddle(self, head, x):
        new_node = Node(x)
        if head is None:
            return new_node

        slow = head
        fast = head

        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        new_node.next = slow.next
        slow.next = new_node

        return head

def printList(head):
    current = head
    values = []
    while current:
        values.append(str(current.data))
        current = current.next
    print(' '.join(values))



arr_str = input().strip()  
x_str = input().strip()    


arr = list(map(int, arr_str.split())) if arr_str else []
x = int(x_str)


if not arr:
    head = None
else:
    head = Node(arr[0])
    current = head
    for val in arr[1:]:
        current.next = Node(val)
        current = current.next


sol = Solution()
head = sol.insertInMiddle(head, x)


printList(head)
